### PR TITLE
feat: add collection feature requests

### PR DESCRIPTION
**Pull request by adam-sroka**
Implement feature requests to the Collection module. These range from restricting household id length, to gps location fetching or page navigation. Definitely room for improvement so I'm happy to either implement requested changes to this PR, or document TODOs for later refactoring-focused PR.

### CODE REVIEW

File: lib/collection/widgets/first_pass_screen.dart [hunk 1]
```diff
@@ -60,7 +60,7 @@ class FirstPassScreen extends StatelessWidget {
                 children: <Widget>[
                   FloatingActionButton.extended(
                       heroTag: null,
-                      label: const Text("New Food Item"),
```
Comment (sshazril) [inline lib/collection/widgets/first_pass_screen.dart:63]: I was under the impression that food item was a specific word used by ICRISAT for this purpose - if we're moving away from using food item then internal variables ought to be updated too? Theres at least one other place in the UI where 'food item' is still used - on the probes screen when accessing via collection

File: lib/collection/widgets/fourth_pass_food_item_card.dart [hunk 2]
```diff
@@ -25,6 +31,8 @@ class FourthPassFoodItemCard extends StatelessWidget {
                 foodItem.timePeriod.value),
             TextFormField(
               readOnly: true,
+              // TODO add snackbar confirmation and scroll in the given page
+              onTap: () => onSelectedScreenChanged!(2),
```
Comment (sshazril) [inline lib/collection/widgets/fourth_pass_food_item_card.dart:35]: With relation to previous conversations about named strings, is the plan to eventually replace these indices with readable strings? (in the sense that '2' as a selected screen is vague without context)

File: lib/collection/widgets/second_pass_food_item_card.dart [hunk 3]
```diff
@@ -34,7 +34,6 @@ class SecondPassFoodItemCard extends StatelessWidget {
       DropdownMenuItem(child: Text('Wild food'), value: 'Wild food'),
       DropdownMenuItem(child: Text('Food aid'), value: 'Food aid'),
       DropdownMenuItem(child: Text('Other'), value: 'Other'),
```
Comment (sshazril) [inline lib/collection/widgets/second_pass_food_item_card.dart:36]: Might be worth adding TODOs to replace all the old dropdownmenu implementations if that is the plan

File: lib/collection/bloc/collection_bloc.dart [hunk 4]
```diff
@@ -389,4 +519,72 @@ class CollectionBloc extends Bloc<CollectionEvent, CollectionState> {
...
+    emit(state.copyWith(gibsonsForm: changedGibsonsForm));
+
+    try {
+      Position position = await _determinePosition();
+
+      geoLocationFormatted =
+          position.latitude.toString() + ', ' + position.longitude.toString();
+    } catch (e) {
+      geoLocationFormatted = 'Undetermined';
+    }
```
[the remark this answers is not in the document]
Comment (adam-sroka) [reply lib/collection/bloc/collection_bloc.dart]: Good suggestion, I implemented it now.

File: lib/collection/widgets/first_pass_food_item_card.dart [hunk 5]
```diff
@@ -65,6 +65,7 @@ class FirstPassFoodItemCard extends StatelessWidget {
             // TODO: Make a pop-up with a confirmation box
             // for deleting an item, this might need to be done in First Pass
             // Screen
+            // Or, make a three dot button that opens a window to delete
```
[the remark this answers is not in the document]
Comment (adam-sroka) [reply lib/collection/widgets/first_pass_food_item_card.dart]: I know that `Slidable` is used in the measurements and in the `Recipe` module, I just made that TODO because I think it is a matter of preference what item deletion method to use — to me, sliding feels like a very iOS thing to do, on Android it is often done either by long pressing or by some visible button, usually something subtle, yet visible, like a three-dot button in the corner of the item. I personally like when the user sees all the available actions, hence I'm a fan of buttons over long-presses or sliding, as these are not visible.

So tl;dr I implemented sliding to delete for now, but I'd be happy if we could run an AB test in the future with different options to see what the enumerators prefer the most 😊.

File: lib/collection/widgets/first_pass_screen.dart (hunk 1 above)
Comment (adam-sroka) [reply lib/collection/widgets/first_pass_screen.dart:63]: I wasn't aware that `Food Item` was a specific word used by ICRISAT — if I remember correctly, I just first used it in Collection because of a lack of better term for the given object. However, the name was inspired to keep the backend expressive, not with the UI in mind.

Again, this just comes down to preference — I personally imagine a user would find it more natural to "add new food" compared to "add new food _item_", and I think it is perfectly alright to keep the user-visible-button name different to the "backend" implementation naming, since the event the button sends is still clear in that it is a FoodItem.

tl;dr we can ask ICRISAT about it.

File: lib/collection/widgets/fourth_pass_food_item_card.dart [hunk 6]
```diff
@@ -118,7 +160,10 @@ class FourthPassFoodItemCard extends StatelessWidget {
                 onPressed: () => onConfirmationChanged!(!foodItem.confirmed),
                 child: foodItem.confirmed
                     ? const Text('Unconfirm')
-                    : const Text('Confirm'))
+                    : const Text('Confirm')),
+            const Divider(),
+            TextButton(
+                onPressed: () => onDeleted!(), child: const Text('Delete'))
```
[the remark this answers is not in the document]
Comment (adam-sroka) [reply lib/collection/widgets/fourth_pass_food_item_card.dart]: Yup, similarly to above, I've implemented `Slidable`, but I have a bit different vie on the information clutter on the screen. I think it would be nice to try out a small button in the top-right corner of each `FoodItem` card for deletion — obviously the current `Delete` button at the bottom was horrendous 😂, but that was just a proof of concept.

We can look into that in the future.

File: lib/collection/widgets/fourth_pass_food_item_card.dart (hunk 2 above)
Comment (adam-sroka) [reply lib/collection/widgets/fourth_pass_food_item_card.dart]: True, done 😊.

File: lib/collection/widgets/third_pass_screen.dart [hunk 7]
```diff
@@ -35,25 +35,44 @@ class ThirdPassScreen extends StatelessWidget {
...
-                                  foodItemMeasurementUnit:
-                                      changedMeasurementUnit)));
+                    foodItem: state.gibsonsForm.foodItems[index],
+                    onMeasurementMethodChanged:
+                        (changedMeasurementIndexAndMethod) => context
+                            .read<CollectionBloc>()
+                            .add(FoodItemMeasurementMethodChanged(
+                                foodItem: state.gibsonsForm.foodItems[index],
+                                measurementIndex:
+                                    changedMeasurementIndexAndMethod[0],
```
[the remark this answers is not in the document]
Comment (adam-sroka) [reply lib/collection/widgets/third_pass_screen.dart]: The `TODO`s to tackle this were already in the `third_pass_food_item_card.dart` — `// TODO: refactor to Map<String, dynamic> and name the arguments as keys` — but I now implemented it.

File: lib/collection/widgets/second_pass_food_item_card.dart (hunk 3 above)
Comment (adam-sroka) [reply lib/collection/widgets/second_pass_food_item_card.dart:36]: Yep, I think this is basically Issue #15 — I made a comment about it there as I think the solution is to replace `DropdownMenuItems` by `DropdownSearch` — I think it now makes more sense to track it as an Issue as there are quite many places where `DropdownMenuItem` is used in the code.

File: lib/collection/widgets/sensitization_screen.dart [hunk 8]
```diff
@@ -117,30 +119,31 @@ class RespondentNameInput extends StatelessWidget {
...
 
   @override
   Widget build(BuildContext context) {
     return BlocBuilder<CollectionBloc, CollectionState>(
       builder: (context, state) {
-        return TextFormField(
-          initialValue: state.gibsonsForm.respondentTelNumber.value,
-          keyboardType: TextInputType.phone,
+        return IntlPhoneField(
+          invalidNumberMessage: 'Enter a valid tel. number',
```
[the remark this answers is not in the document]
Comment (adam-sroka) [reply lib/collection/widgets/sensitization_screen.dart:130]: Yes, I noticed this one as well and tried to fix it before. The problem is that it is a bug in the library that we're using as far as I can tell, and all other libraries I tried did not work at all.

I'll open up an issue about this, and first see if I can fix this using the API of the library — there seems to be a `validator` method that can be supplied, but I don't think the problem is that the validation is not working, rather that it doesn't update upon switching of the country. If I cannot fix it with their API, I'll look if I could make changes and submit a PR to incorporate into the library.

[the remark this answers is not in the document]
Comment (adam-sroka) [reply lib/collection/widgets/sensitization_screen.dart:130]: Tracked in #27

**Comment by adam-sroka** [comment]
I hopefully acted on all of the requested changes, either implementing them or tracking them as a `TODO` or an issue. Lmk how it seems 😊.
